src/db.c

```diff
@@ -912,15 +912,71 @@ void keysCommand(client *c) {
     setDeferredArrayLen(c, replylen, numkeys);
 }
 
+#define DEFAULT_SCAN_COMMAND_COUNT 10
+
+/* The SCAN command's default COUNT is 10.
+ * Since it may store keys + values, the
+ * buffer size is roughly 10 * 2 = 20.
+ * Adding a 20% buffer (20 * 1.2) gives 24. */
+#define MAX_SCAN_ARRAY_BUFFER 24
+
+/* Used to store the iteration results from the SCAN/HSCAN/SSCAN/ZSCAN commands. */
+typedef struct {
+    sds *array;  /* Points to buf or heap */
+    size_t used; /* Used size of the array */
+    size_t size; /* Allocated size of the array */
+    void (*free)(void *);
+    sds buf[MAX_SCAN_ARRAY_BUFFER]; /* Pre-allocated buffer reduces heap allocation */
+} scanArray;
+
+/* Initialize a scanArray structure. */
+static void scanArrayInit(scanArray *result, void (*free)(void *)) {
+    /* The array is initialized with a pre-allocated buffer of size MAX_SCAN_ARRAY_BUFFER,
+     * which will be used until the number of elements exceeds this size. */
+    result->used = 0;
+    result->array = result->buf;
+    result->size = MAX_SCAN_ARRAY_BUFFER;
+
+    /* The free function is used to free the elements in the array,
+     * if NULL then no free will be performed. */
+    result->free = free;
+}
+
+/* Append a value to the scanArray. */
+static void scanArrayAppend(scanArray *result, sds v) {
+    if (result->used == result->size) {
+        result->size *= 2;
+        if (result->array == result->buf) {
+            result->array = zmalloc(sizeof(sds) * result->size);
+            memcpy(result->array, result->buf, sizeof(result->buf));
+        } else {
+            result->array = zrealloc(result->array, sizeof(sds) * result->size);
+        }
+    }
+    result->array[result->used++] = v;
+}
+
+/* Clean a scanArray inited with scanArrayInit(). */
+static void scanArrayCleanup(scanArray *result) {
+    if (result->free) {
+        for (size_t i = 0; i < result->used; i++) {
+            result->free(result->array[i]);
+        }
+    }
+    if (result->buf != result->array) {
+        zfree(result->array);
+    }
+}
+
 /* Data used by the dict scan callback. */
 typedef struct {
-    list *keys;     /* elements that collect from dict */
-    robj *o;        /* o must be a hash/set/zset object, NULL means current db */
-    serverDb *db;   /* database currently being scanned */
-    long long type; /* the particular type when scan the db */
-    sds pattern;    /* pattern string, NULL means no pattern */
-    long sampled;   /* cumulative number of keys sampled */
-    int only_keys;  /* set to 1 means to return keys only */
+    scanArray *result; /* elements that collect from dict */
```

Comment:
Should we put the struct inline in the scanData struct to save another allocation?

```suggestion
    scanArray result; /* elements that collect from dict */
```

Reply:
`scanArray *result` is a pointer to a stack object and will not cause memory allocation, so it is better to use `scanArray *result`

Reply:
True, but scanData is also stack-allocated. We could save 8 bytes to avoid storing a pointer but it doesn't matter. Simplicity and clarity is more important.